pkg/cnab/config-adapter/adapter.go

```diff
@@ -166,134 +166,151 @@ func (c *ManifestConverter) generateDefaultAction(action string) bundle.Action {
+func (c *ManifestConverter) generateParameterDefinition(ctx context.Context, defs *definition.Definitions,
+	sourceParam manifest.ParameterDefinition) bundle.Parameter {
 	log := tracing.LoggerFromContext(ctx)
 
-	outputs := make(map[string]bundle.Output, len(c.Manifest.Outputs))
+	// Update ApplyTo per parameter definition and manifest
+	sourceParam.UpdateApplyTo(c.Manifest)
 
-	addOutput := func(output manifest.OutputDefinition) {
-		o := bundle.Output{
-			Definition:  output.Name,
-			Description: output.Description,
-			ApplyTo:     output.ApplyTo,
-			// must be a standard Unix path as this will be inside of the container
-			// (only linux containers supported currently)
-			Path: path.Join(config.BundleOutputsDir, output.Name),
-		}
+	p := bundle.Parameter{
+		Definition:  sourceParam.Name,
+		ApplyTo:     sourceParam.ApplyTo,
+		Description: sourceParam.Description,
+	}
 
-		if output.Sensitive {
-			output.Schema.WriteOnly = toBool(true)
+	// If the default is empty, set required to true.
+	// State parameters are always optional, and don't have a default
+	if sourceParam.Default == nil && !sourceParam.IsState {
+		p.Required = true
+	}
+
+	if sourceParam.Sensitive {
+		sourceParam.Schema.WriteOnly = toBool(true)
+	}
+
+	if !sourceParam.Destination.IsEmpty() {
+		p.Destination = &bundle.Location{
+			EnvironmentVariable: sourceParam.Destination.EnvironmentVariable,
+			Path:                manifest.ResolvePath(sourceParam.Destination.Path),
+		}
+	} else {
+		p.Destination = &bundle.Location{
+			EnvironmentVariable: manifest.ParamToEnvVar(sourceParam.Name),
 		}
+	}
 
-		if output.Type == nil {
-			// Default to a file type if the param is stored in a file
-			if output.Path != "" {
-				output.Type = "file"
-			} else {
-				// Assume it's a string otherwise
-				output.Type = "string"
-			}
-			log.Debugf("Defaulting the type of output %s to %s", output.Name, output.Type)
+	if sourceParam.Type == nil {
+		// Default to a file type if the param is stored in a file
+		if sourceParam.Destination.Path != "" {
+			sourceParam.Type = "file"
+		} else {
+			// Assume it's a string otherwise
+			sourceParam.Type = "string"
 		}
 
-		// Create a definition that matches the output if one isn't already defined
-		if _, ok := (*defs)[output.Name]; !ok {
-			kind := "output"
-			if output.IsState {
-				kind = "state"
-			}
+		log.Debugf("Defaulting the type of parameter %s to %s", sourceParam.Name, sourceParam.Type)
+	}
 
-			defName := c.addDefinition(output.Name, kind, output.Schema, defs)
-			o.Definition = defName
+	// Create a definition that matches the parameter if one isn't already defined
+	if _, ok := (*defs)[sourceParam.Name]; !ok {
+		kind := "parameter"
+		if sourceParam.IsState {
+			kind = "state"
 		}
-		outputs[output.Name] = o
-	}
 
-	for _, o := range c.Manifest.Outputs {
-		addOutput(o)
+		defName := c.addDefinition(sourceParam.Name, kind, sourceParam.Schema, defs)
```

Comment:
```suggestion
		p.Definition = c.addDefinition(sourceParam.Name, kind, sourceParam.Schema, defs)
```

Reply:
Right now I believe most of our stuff goes to stdOut, I would love to implement a structured logger so we could follow the span throughout multiple functions, appending more data, and being able to see exactly what func errors. 
I believe right now because we abstract away our logging that implementing a structured logger wouldn't be a huge undertaking.

Reply:
Oh hey, this is new https://go.dev/blog/slog

Reply:
Structured log is good, but tracing is better. Tracing is structured logs + causality through call hierarchy. This is probably going to be more apparent in the operator as stuff will be less going to stdout and more to some telemetry ingestion pipeline.